crates/nu-path/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+    let path = match std::fs::read_link(&absolutized) {
+        Ok(resolved) => {
+            let parent = absolutized.parent().unwrap_or(&absolutized);
+            absolutize(parent, resolved)
+        }
+
+        Err(e) => {
+            if absolutized.exists() {
+                absolutized
+            } else {
+                return Err(e);
+            }
+        }
+    };
+
+    Ok(dunce::simplified(&path).to_path_buf())
+}
+
+// Expansion logic lives here to enable testing without depending on dirs-next
+fn expand_tilde_with<'a>(path: Cow<'a, Path>, home: Option<PathBuf>) -> Cow<'a, Path> {
+    if !path.starts_with("~") {
+        return path;
+    }
+
+    match home {
+        None => path,
+        Some(mut h) => {
+            if h == Path::new("/") {
+                // Corner case: `h` root directory;
+                // don't prepend extra `/`, just drop the tilde.
+                cow_map_by_ref(path, |p: &Path| {
+                    p.strip_prefix("~").expect("cannot strip ~ prefix")
+                })
+            } else {
+                h.push(path.strip_prefix("~/").expect("cannot strip ~/ prefix"));
+                Cow::Owned(h)
+            }
+        }
+    }
+}
+
+pub fn expand_tilde<'a>(path: Cow<'a, Path>) -> Cow<'a, Path> {
+    expand_tilde_with(path, dirs_next::home_dir())
+}
+
+pub fn expand_tilde_string<'a>(path: Cow<'a, str>) -> Cow<'a, str> {
+    cow_map_str_path(path, expand_tilde)
+}
+
+
+
+// Remove "." and ".." in a path. Prefix ".." are not removed as we don't have access to the
+// current dir. This is merely 'string manipulation'. Does not handle "...+", see expand_ndots for that
+pub fn resolve_dots<'a>(path: Cow<'a, Path>) -> Cow<'a, Path> {
+    debug_assert!(!path.components().any(|c| std::matches!(c, Component::Normal(os_str) if os_str.to_string_lossy().starts_with("..."))), "Unexpected ndots!");
+    if !path
+        .components()
+        .any(|c| std::matches!(c, Component::CurDir | Component::ParentDir))
+    {
+        return path;
+    }
+
+    let mut result = PathBuf::with_capacity(path.as_os_str().len());
+
+    // Only pop/skip path elements if the previous one was an actual path element
+    let prev_is_normal = |p: &Path| -> bool {
+        p.components()
+            .next_back()
+            .map(|c| std::matches!(c, Component::Normal(_)))
+            .unwrap_or(false)
+    };
+    path.as_ref()
+        .components()
+        .for_each(|component| match component {
+            Component::ParentDir if prev_is_normal(&result) => {
+                result.pop();
+            }
+            Component::CurDir if prev_is_normal(&result) => {}
+            _ => result.push(component),
+        });
+
+    Cow::Owned(dunce::simplified(&result).to_path_buf())
+}
+
+// Expands ~ to home and shortens paths by removing unecessary ".." and "."
+// where possible. Also expands "...+" appropriately.
+pub fn expand_path<'a>(path: Cow<'a, Path>) -> Cow<'a, Path> {
```

Comment:
For a while most public functions in this file has the signature
fn(p: impl Into<Cow<'a, Path>>) -> Cow<'a, Path>
which allows it to be called with fn(&Path) or fn(PathBuf).
I removed them because the signatures started becoming fairly complex and on the callside it was a bit magical what you were actually passing as an arg. I can switch back to it if you prefer simplifying the call sites or even &Path if Cow is too unweildy.